### PR TITLE
WYSIWYG Improvements to highlight

### PR DESCRIPTION
**Pull request by chlafreniere**
Fixes highlight portion of #12467, which is feedback that's been echoed another time.

Needed to fix the following:
- Escaping text with < or > in them (otherwise unhighlight wouldn't work properly)
- Highlight/unhighlight when it's wrapped either in a mark or span (with background color defined)

Tests updated and new tests added.

### CODE REVIEW

**Comment by lucyzhang929** [comment]
When I click the highlight button again in WYSIWYG mode, it still doesn’t undo the action. It works if I highlight something in markdown mode and switch back to WYSIWYG mode - then clicking the highlight button will undo the highlight.

**Comment by Charles-Gagnon** [comment]
This is probably something to work on separately - but does this handle highlighted text in the middle of a string? i.e. I have some words with one of them highlighted. I selected all the words and hit highlight - it highlights the entire selected string. I then hit highlight again - does it remove the highlight from all the strings?

**Comment by chlafreniere** [comment]
@lucyzhang929 thanks. When exiting the cell, the following occurs:

```
<p><span style="background-color: yellow;">Highlight me</span>&nbsp;</p>
```
gets converted into
```
<p><mark>Highlight me</mark></p>
```
because we're sending the following to be rendered from markdown:
```
<mark>Highlight me</mark>
```



So that's why it works in some cases (exiting WYSIWYG) rather than others.

Why is highlight the weird one? It's the case where executing the same command again doesn't actually undo the highlight; e.g. for bold, executing the bold command on text that's already bold undoes the bold. So, we're stuck implementing this ourselves.

We could implement some logic to look for `mark` or `style="background-color: yellow;"` and if either of these are found, remove them. I don't love this solution to be honest, but if we have to, we have to.

Alternatively, we could force an update to WYSIWYG preview on every keystroke, but we'd have to be very cognizant of potential perf issues here.

Don't love either option to be honest. Any ideas?

**Review by halerankin — APPROVED** [review]

File: src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts [hunk 1]
```diff
@@ -174,19 +178,19 @@ suite('HTML Markdown Converter', function (): void {
...
 		htmlString = '<ul><li>&lt;test&gt;</li></ul>';
 		assert.equal(htmlMarkdownConverter.convert(htmlString), '- \\<test\\>', 'Basic unordered list non-HTML tag item test failed to escape');
 		htmlString = '<ol><li>&lt;test&gt;</li></ol>';
 		assert.equal(htmlMarkdownConverter.convert(htmlString), '1. \\<test\\>', 'Basic ordered list non-HTML tag item test failed to escape');
 		htmlString = '<mark>&lt;test&gt;</mark>';
-		assert.equal(htmlMarkdownConverter.convert(htmlString), '<mark>&lt;test&gt;</mark>', 'Basic highlighting Non-HTML tag test failed to escape');
+		assert.equal(htmlMarkdownConverter.convert(htmlString), '<mark>\\<test\\></mark>', 'Basic highlighting Non-HTML tag test failed to escape');
 		htmlString = '<mark><h1>&lt;test&gt;</h1></mark>';
-		assert.equal(htmlMarkdownConverter.convert(htmlString), '<mark><h1>&lt;test&gt;</h1></mark>', 'Non-HTML tag inside multiple html tags test failed to escape');
+		assert.equal(htmlMarkdownConverter.convert(htmlString), '<mark>\n\n# \\<test\\>\n\n</mark>', 'Non-HTML tag inside multiple html tags test failed to escape');
```
Comment (chlafreniere) [inline src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts:189]: Notice that we're now doing the proper conversion to markdown

**Comment by chlafreniere** [comment]
> When I click the highlight button again in WYSIWYG mode, it still doesn’t undo the action. It works if I highlight something in markdown mode and switch back to WYSIWYG mode - then clicking the highlight button will undo the highlight.

I believe I've fixed this now.

**Comment by chlafreniere** [comment]
@VasuBhog I've changed some of the escaping logic here and think that we're more resilient now. Can you take a look please? Especially at the diffs in the tests, if you could.

File: src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts (hunk 1 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts:189]: Why do we add two new lines with <mark> tag?

Comment (chlafreniere) [reply src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts:189]: Headings always get converted like that. Here's a code snippet from our heading rule:

```
				if (options.headingStyle === 'setext' && hLevel < 3) {
					let underline = '#'.repeat(hLevel);
					return '\n\n' + content + '\n' + underline + '\n\n';
				} else {
					return '\n\n' + '#'.repeat(hLevel) + ' ' + content + '\n\n';
				}
```

File: src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts [hunk 2]
```diff
@@ -100,8 +137,6 @@ export class HTMLMarkdownConverter {
 				const notebookLink = node.href ? URI.parse(node.href) : URI.file(node.title);
 				const notebookFolder = this.notebookUri ? path.join(path.dirname(this.notebookUri.fsPath), path.sep) : '';
 				let relativePath = findPathRelativeToContent(notebookFolder, notebookLink);
-				node.innerText = escapeAngleBrackets(node.innerText);
-				content = escapeAngleBrackets(content);
```
Comment (corivera) [inline src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:104]: Why is this escaping being removed?

Comment (chlafreniere) [reply src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:104]: It wasn't working properly (not handling nested elements correctly). So, if I had the following:

```
<mark>He**llo** world</mark>
```

The unhighlight result would be
```Hello world```

Where instead, it should be

```He**llo** world```

Instead, now leveraging the text escaping mechanism.

**Comment by barbaravaldez** [comment]
I tested some scenarios and undoing the highlight in WYSIWYG does not work in nested elements.
Screen Shot 2020-12-02 at 1 00 13 PM

File: src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts [hunk 3]
```diff
@@ -65,7 +65,44 @@ export class TransformMarkdownAction extends Action {
 				document.execCommand('formatBlock', false, 'H3');
 				break;
 			case MarkdownButtonType.HIGHLIGHT:
-				document.execCommand('hiliteColor', false, 'Yellow');
+				let selectionFocusNode = document.getSelection()?.focusNode;
+				// Find if element is wrapped in <mark></mark>
+				while (selectionFocusNode?.parentNode?.nodeName?.toLowerCase() && selectionFocusNode?.parentNode?.nodeName?.toLowerCase() !== 'mark') {
+					selectionFocusNode = selectionFocusNode.parentNode;
+				}
+				// Find if element is wrapped in <span background-color="yellow"
```
Comment (corivera) [inline src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts:73]: Nit: ">" at the end of this span tag in the comment

**Comment by chlafreniere** [comment]
> I tested some scenarios and undoing the highlight in WYSIWYG does not work in nested elements.
> Screen Shot 2020-12-02 at 1 00 13 PM

I'm a bit confused. Can you give me some repro steps? I tried these and they seemed to work for me so just trying to understand where I've gone wrong 😄 .

Example:

WYSIWYG:
image

WYSIWYG + highlight:
image

WYSIWYG + unhighlight:
image

Split view:
image

**Review by corivera — APPROVED** [review]

**Comment by barbaravaldez** [comment]
> > I tested some scenarios and undoing the highlight in WYSIWYG does not work in nested elements.
> > Screen Shot 2020-12-02 at 1 00 13 PM
> 
> @barbaravaldez I'm a bit confused. Can you give me some repro steps? I tried these and they seemed to work for me so just trying to understand where I've gone wrong 😄 .
> 
> Example:
> 
> WYSIWYG:
> image
> 
> WYSIWYG + highlight:
> image
> 
> WYSIWYG + unhighlight:
> image
> 
> Split view:
> image

@chlafreniere I think it only happens when `<span><hello><span>` is underlined 😮 The first time I press the highlight button it gets highlighted, but after clicking multiple times, it doesn't undo the action. When I exit the cell and try again, it works.

**Review by lucyzhang929 — APPROVED** [review]
I'm also seeing the same issue @barbaravaldez mentioned (underline, highlight, highlight again does not un-highlight). But clicking underline again and then highlight will successfully undo both.

**Review by barbaravaldez — APPROVED** [review]

**Review by VasuBhog — APPROVED** [review]
Tested this against WYSIWYG tags notebook seems to maintain consistency as well as improve highlighting.

**Comment by halerankin** [comment]
Just tested the Code style. Highlighted a single word with no other styles on it, clicked Code, clicked Code again: 
image
If, after clicking Code the first time, I highlight the tick marks above and below the word and click Code, I get the same results.

**Comment by chlafreniere** [comment]
@halerankin can you create a bug for that? Don't want to feature-creep too much here and stick to highlight as much as I can.

**Comment by chlafreniere** [comment]
@lucyzhang929 @barbaravaldez thanks. Given that there's a workaround let's address this in a subsequent PR as it shouldn't block anything.

**Pull request merged**